numpy/lib/_function_base_impl.py

```diff
@@ -4716,55 +4821,106 @@ def _quantile(
-                                                      values_count)
-        # --- Sorting
-        arr.partition(
-            np.unique(np.concatenate(([0, -1],
-                                      previous_indexes.ravel(),
-                                      next_indexes.ravel(),
-                                      ))),
-            axis=0)
-        if supports_nans:
-            slices_having_nans = np.isnan(arr[-1, ...])
+        # Weighted case, we need to sort anyway.
+        # This implements method="inverted_cdf".
+        weights = np.asanyarray(weights)
+        if axis != 0:
+            weights = np.moveaxis(weights, axis, destination=0)
+        index_array = np.argsort(arr, axis=0, kind="stable")
+        # TODO: Which value to set to slices_having_nans.
+        slices_having_nans = None
+        # TODO: Deal with nan values, e.g. np.isnan(arr(index_array[-1])) by
+        # be true.
+        # TODO: Special case quantiles == 0 and quantiles == 1?
+
+        # arr = arr[index_array, ...]  # but this adds trailing dimensions of
+        # 1.
+        arr = np.take_along_axis(arr, index_array, axis=0)
+        if weights.shape == arr.shape:
+            weights = np.take_along_axis(weights, index_array, axis=0)
         else:
-            slices_having_nans = None
-        # --- Get values from indexes
-        previous = arr[previous_indexes]
-        next = arr[next_indexes]
-        # --- Linear interpolation
-        gamma = _get_gamma(virtual_indexes, previous_indexes, method)
-        result_shape = virtual_indexes.shape + (1,) * (arr.ndim - 1)
-        gamma = gamma.reshape(result_shape)
-        result = _lerp(previous,
-                       next,
-                       gamma,
-                       out=out)
+            # weights is 1d
+            weights = weights.reshape(-1)[index_array, ...]
+        # weights = np.take_along_axis(weights, index_array, axis=axis)
+        # We use the weights to calculate the empirical distribution function.
+        cdf = weights.cumsum(axis=0, dtype=float)
+        cdf /= cdf[-1, ...]  # normalization
+        # Only inverted_cdf is supported. Search index i such that
+        # sum(weights[j], j=0..i-1) < quantile <= sum(weights[j], j=0..i)
+
+        def find_cdf_1d(arr, cdf):
+            i = np.searchsorted(cdf, quantiles, side="left")
+            # We might have reached the maximum with i = len(arr), e.g. for
+            # quantiles == 1.
+            i = np.minimum(i, values_count - 1)
+            result = take(arr, i, axis=0, out=out)
+            return result
+
+        r_shape = arr.shape[1:]
+        if quantiles.ndim > 0: 
+            r_shape += quantiles.shape
+        result = np.empty_like(arr, shape=r_shape)
+        # See apply_along_axis, which we do for axis=0. Note that Ni = (,)
+        # always, so we remove it here.
+        Nk = arr.shape[1:]
+        for kk in np.ndindex(Nk):
+            result[kk + np.s_[..., ]] = find_cdf_1d(
```

Comment:
Would just use a `(...,)` rather than the fancy `s_`.  But yeah, vectorize isn't better...
(The `...` shouldn't be needed at all there?)

Reply:
https://github.com/numpy/numpy/pull/24254/commits/4f7841a907802cfa19d1123c526ae730ad77dafb